File: agent/analyze_ticker.py
"""On-demand single-ticker swing analysis (BUY / WAIT / PASS) -- the engine
behind the planned Telegram /add <ticker> watchlist command.

Runs the same core pieces as the daily scan (market regime, technicals,
chart-vision grading, options walls, confluence) against ONE ticker and
applies the verdict rules in strategy/VERDICT_RULES.md. Self-contained: no
Claude Code agentic session needed -- chart_vision.grade() already calls the
Anthropic API directly (ANTHROPIC_API_KEY from env), same as the daily
routine's own chart-grading step.

Two things this does NOT try to replicate from the daily pipeline, both
noted on the returned Verdict:
  - RS percentile is normally ranked against that day's full screener pool.
    A single ad-hoc ticker has no pool, so `rs_pctile_is_estimate=True` marks
    a proxy: how far the ticker's weighted RS return beats SPY's own, scaled
    onto a 0-100 range calibrated so "beats SPY by 15pp" ~= 85th percentile.
    Real peer-pool ranking (e.g. against sector-rotation's cached S&P 500
    prices) is a fair v2 upgrade, not done here.
  - No halal screen -- same standing override as the daily scan.

Usage:
    python -m agent.analyze_ticker NVDA
    python -m agent.analyze_ticker NVDA --options-token <MARKETDATA_API_TOKEN>
"""
from __future__ import annotations
import argparse
import json
import sys
from dataclasses import asdict, dataclass, field

from . import data, market_regime, options_walls, sector_rotation, technicals
from . import chart_vision

# yfinance's sector taxonomy doesn't match the 11 GICS SPDR sector names
# 1:1 (e.g. "Basic Materials" vs "Materials") -- map the ones that differ.
YF_SECTOR_MAP = {
    "Technology": "Technology",
    "Consumer Cyclical": "Consumer Discretionary",
    "Financial Services": "Financials",
    "Industrials": "Industrials",
    "Energy": "Energy",
    "Healthcare": "Health Care",
    "Consumer Defensive": "Consumer Staples",
    "Utilities": "Utilities",
    "Real Estate": "Real Estate",
    "Basic Materials": "Materials",
    "Communication Services": "Communication Services",
}

# --- verdict rule thresholds, see strategy/VERDICT_RULES.md ---
MIN_RR = 1.0
MIN_EARNINGS_TRADING_DAYS = 7
MAX_BASE_NUMBER = 4              # base_number >= this -> PASS
APLUS_WAIT_CEILING = 6           # aplus_score <= this -> WAIT (also caps BUY at >=7)
BREAKOUT_RVOL_MIN = 1.2          # trigger: breakout day needs at least this much relative volume
RS_PROXY_SPY_MARGIN_FOR_85TH = 0.15  # beating SPY's weighted return by this much ~= 85th pctile proxy


@dataclass
class Verdict:
    ticker: str
    verdict: str                     # "BUY" | "WAIT" | "PASS"
    reason: str
    conviction: str | None = None    # "A+" | "standard" (BUY only)
    rr_band: str | None = None       # "2:1-2.9:1" | "3:1+" (BUY only)
    trigger: str | None = None       # "breakout" | "pullback" | None
    aplus_score: int = 0
    aplus_detail: list = field(default_factory=list)
    regime_score: int = 0
    regime_mode: str = ""
    sector: str | None = None
    sector_beats_spy: bool | None = None
    price: float | None = None
    entry: float | None = None
    stop: float | None = None
    target: float | None = None
    rr: float | None = None
    confluence_count: int = 0
    confluence_signals: list = field(default_factory=list)
    chart_grade: str | None = None
    vision_note: str | None = None
    rs_pctile_estimate: float | None = None
    rs_pctile_is_estimate: bool = True
    earnings_trading_days: float | None = None
    options_wall_source: str | None = None
    chart_path: str | None = None


def _trading_days(calendar_days: float) -> float:
    return calendar_days * 5 / 7


def _sector_and_beats_spy(ticker: str, sectors_df) -> tuple[str | None, bool]:
    info = data.info(ticker)
    yf_sector = info.get("sector")
    mapped = YF_SECTOR_MAP.get(yf_sector, yf_sector)
    if not mapped or sectors_df.empty:
        return mapped, False
    row = sectors_df[sectors_df["sector"] == mapped]
    if row.empty:
        return mapped, False
    return mapped, bool(row.iloc[0]["4W_vs_SPY"] > 0)


def _rs_pctile_estimate(t) -> float | None:
    """Proxy for a peer-pool RS percentile: how far this ticker's weighted RS
    return beats SPY's own, mapped onto ~0-100 so a 15pp margin lands near 85.
    See module docstring -- not a real percentile."""
    if t.rs_raw != t.rs_raw:  # nan
        return None
    spy_df = data.history("SPY", period="2y")
    if spy_df.empty:
        return None
    spy_rs = technicals.rs_weighted_return(spy_df["Close"])
    if spy_rs != spy_rs:
        return None
    margin = t.rs_raw - spy_rs
    pctile = 50 + (margin / RS_PROXY_SPY_MARGIN_FOR_85TH) * 35
    return round(min(max(pctile, 0), 100), 1)


def _aplus_checklist(regime_score: int, sector_beats_spy: bool, t, vision: dict,
                      rr: float | None, rs_pctile: float | None,
                      earnings_trading_days: float | None) -> tuple[int, list]:
    """Same 9-question checklist as technicals.a_plus_score(), but with the
    earnings item pointed at MIN_EARNINGS_TRADING_DAYS (this tool's own
    agreed threshold) instead of that function's hardcoded 14-calendar-day
    default, and rs_pctile passed in explicitly (the estimate above, not a
    peer-pool rank)."""
    checks = [
        ("Regime 3/4 or 4/4", regime_score >= 3),
        ("Sector beats SPY (4W)", sector_beats_spy),
        ("RS percentile >= 85 (estimate)", (rs_pctile or 0) >= 85),
        ("Recognizable pattern", vision.get("pattern") not in (None, "none")),
        ("VDU present", bool(t.vdu or vision.get("vdu"))),
        ("Clear pivot", t.pivot is not None and not t.extended),
        ("Logical stop", True),
        ("R/R >= 2:1", rr is not None and rr >= 2.0),
        (f"Earnings >= {MIN_EARNINGS_TRADING_DAYS} trading days away",
         earnings_trading_days is None or earnings_trading_days >= MIN_EARNINGS_TRADING_DAYS),
    ]
    return sum(1 for _, ok in checks if ok), [{"check": c, "ok": ok} for c, ok in checks]


def _detect_trigger(t, vision: dict) -> str | None:
    """Breakout: at/through pivot on RVOL>=BREAKOUT_RVOL_MIN, not extended.
    Pullback: VDU (numeric or vision-confirmed) -- assumed sitting at a level
    since the WAIT gate above already requires confluence_count>=1 to reach
    here. Neither -> no trigger today, stays WAIT."""
    if t.pivot is not None and t.dist_to_pivot_pct is not None:
        if t.dist_to_pivot_pct >= -1.0 and not t.extended and t.vol_ratio >= BREAKOUT_RVOL_MIN:
            return "breakout"
    if t.vdu or vision.get("vdu"):
        return "pullback"
    return None


def analyze(ticker: str, options_token: str | None = None) -> Verdict:
    ticker = ticker.upper().strip()
    df = data.history(ticker, period="2y")
    if df.empty:
        return Verdict(ticker=ticker, verdict="PASS", reason="no price data from yfinance")

    t = technicals.read(df)
    regime = market_regime.assess()
    sectors_df = sector_rotation.score()
    sector, sector_beats_spy = _sector_and_beats_spy(ticker, sectors_df)

    chart_path = chart_vision.render_chart(ticker, df, t=t)
    vision = chart_vision.grade(ticker, chart_path, context={
        "price": t.price, "ema9": t.ema9, "ema21": t.ema21, "ema50": t.ema50, "ema200": t.ema200,
        "rsi14": t.rsi14, "atr_pct": t.atr_pct, "pivot": t.pivot,
        "dist_to_pivot_pct": t.dist_to_pivot_pct, "vol_ratio": t.vol_ratio, "vdu": t.vdu,
    })
    chart_grade = vision.get("grade", "C")

    earnings_calendar_days = data.next_earnings_days(ticker)
    earnings_trading_days = (
        _trading_days(earnings_calendar_days) if earnings_calendar_days is not None else None
    )

    entry = vision.get("entry") or t.price
    stop = vision.get("stop")
    target = vision.get("target")
    rr = round((target - entry) / (entry - stop), 2) if (stop and target and entry > stop) else None

    walls = options_walls.read(ticker, t.price, token=options_token)
    conf_count, conf_signals = technicals.confluence_count(t, entry, walls)

    rs_pctile = _rs_pctile_estimate(t)

    v = Verdict(
        ticker=ticker, verdict="WAIT", reason="",
        regime_score=regime.score, regime_mode=regime.mode,
        sector=sector, sector_beats_spy=sector_beats_spy,
        price=t.price, entry=entry, stop=stop, target=target, rr=rr,
        confluence_count=conf_count, confluence_signals=conf_signals,
        chart_grade=chart_grade, vision_note=vision.get("note"),
        rs_pctile_estimate=rs_pctile,
        earnings_trading_days=earnings_trading_days,
        options_wall_source=walls.source,
        chart_path=str(chart_path),
    )

    # ---- PASS gates (any one -> stop here) ----
    if not t.above_200:
        v.verdict, v.reason = "PASS", "below EMA200 -- no long structure"
        return v
    if rr is None or rr < MIN_RR:
        v.verdict, v.reason = "PASS", f"R/R {rr if rr is not None else 'n/a'} < {MIN_RR}"
        return v
    if chart_grade == "F":
        v.verdict, v.reason = "PASS", f"chart grade F -- {vision.get('note', '')}"
        return v
    if earnings_trading_days is not None and earnings_trading_days < MIN_EARNINGS_TRADING_DAYS:
        v.verdict, v.reason = "PASS", f"earnings in ~{earnings_trading_days:.0f} trading days"
        return v
    base_number = vision.get("base_number")
    if base_number is not None and base_number >= MAX_BASE_NUMBER:
        v.verdict, v.reason = "PASS", f"base #{base_number} -- too extended a series of bases"
        return v

    aplus_score, aplus_detail = _aplus_checklist(
        regime.score, sector_beats_spy, t, vision, rr, rs_pctile, earnings_trading_days,
    )
    v.aplus_score, v.aplus_detail = aplus_score, aplus_detail
    trigger = _detect_trigger(t, vision)
    v.trigger = trigger

    # ---- WAIT caps ----
    if regime.score <= 1:
        v.verdict, v.reason = "WAIT", "regime NO_TRADE -- watchlist only"
        return v
    if aplus_score <= APLUS_WAIT_CEILING:
        v.verdict, v.reason = "WAIT", f"A+ score {aplus_score}/9 -- no entry"
        return v
    if t.extended:
        v.verdict, v.reason = "WAIT", f"{t.dist_to_pivot_pct:.1f}% past pivot -- chase risk"
        return v
    if conf_count == 0:
        v.verdict, v.reason = "WAIT", "no confluence at proposed entry"
        return v
    if trigger is None:
        v.verdict, v.reason = "WAIT", "structurally fine, no trigger today"
        return v

    # ---- BUY ----
    v.verdict = "BUY"
    v.conviction = "A+" if aplus_score == 9 else "standard"
    v.rr_band = "3:1+" if rr >= 3.0 else "2:1-2.9:1"
    v.reason = f"{trigger} trigger, A+ {aplus_score}/9, R/R {rr:.1f}"
    return v


def main():
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("ticker")
    parser.add_argument("--options-token", default=None, help="MARKETDATA_API_TOKEN (optional)")
    args = parser.parse_args()
    v = analyze(args.ticker, options_token=args.options_token)
    print(json.dumps(asdict(v), indent=2, default=str))
    print(f"\n{'='*40}\n{v.ticker}: {v.verdict} -- {v.reason}\n{'='*40}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: web/lib/analyzeTicker.ts
/** TypeScript port of agent/analyze_ticker.py -- see strategy/VERDICT_RULES.md
 * for the exact BUY/WAIT/PASS rules. This is the Node/Vercel-reachable
 * version used by the Telegram webhook (web/app/api/telegram-webhook/route.ts)
 * since the Python pipeline isn't part of this deployment (see
 * strategy/TELEGRAM_BOT.md). Same thresholds, same gate order, same
 * confluence/A+ logic -- now includes a real rendered chart image sent to
 * the grading call (lib/renderChart.ts + lib/visionGrade.ts), the same
 * idea as agent/chart_vision.py, just OpenAI instead of Anthropic. */
import { fetchDailyBars, fetchNextEarningsCalendarDays, fetchSectorAndIndustry } from "./marketData";
import { readTechnicals, rsWeightedReturn, confluenceCount, stopAndEntry, type TechRead } from "./technicals";
import { assessRegime } from "./marketRegime";
import { scoreSectors } from "./sectorRotation";
import { readOptionsWalls, type OptionsWalls } from "./optionsWalls";
import { gradeChart, type VisionGrade } from "./visionGrade";
import { renderChartPng } from "./renderChart";
import { mapToTheme } from "./themes";
import { getLatestRrgReading, type Quadrant } from "./rrgQuadrant";

const YF_SECTOR_MAP: Record<string, string> = {
  "Technology": "Technology",
  "Consumer Cyclical": "Consumer Discretionary",
  "Financial Services": "Financials",
  "Industrials": "Industrials",
  "Energy": "Energy",
  "Healthcare": "Health Care",
  "Consumer Defensive": "Consumer Staples",
  "Utilities": "Utilities",
  "Real Estate": "Real Estate",
  "Basic Materials": "Materials",
  "Communication Services": "Communication Services",
};

export const MIN_RR = 1.0;
export const MIN_EARNINGS_TRADING_DAYS = 7;
export const MAX_BASE_NUMBER = 4;
export const APLUS_WAIT_CEILING = 6;
export const BREAKOUT_RVOL_MIN = 1.2;
export const RS_PROXY_SPY_MARGIN_FOR_85TH = 0.15;

export type VerdictType = "BUY" | "WAIT" | "PASS";

export interface Verdict {
  ticker: string;
  verdict: VerdictType;
  reason: string;
  conviction: "A+" | "A" | "standard" | null;
  rr_band: "2:1-2.9:1" | "3:1+" | null;
  trigger: "breakout" | "pullback" | null;
  aplus_score: number;
  aplus_detail: { check: string; ok: boolean }[];
  regime_score: number;
  regime_mode: string;
  sector: string | null;
  sector_beats_spy: boolean;
  sector_rrg_quadrant: Quadrant | null;
  theme: string | null;
  theme_match_confidence: "exact" | "approximate" | null;
  theme_rrg_quadrant: Quadrant | null;
  price: number | null;
  entry: number | null;
  stop: number | null;
  target: number | null;
  rr: number | null;
  confluence_count: number;
  confluence_signals: string[];
  chart_grade: string | null;
  vision_note: string | null;
  /** 3-4 sentence plain-language explanation (what the setup looks like,
   * why it isn't a buy yet, and what to watch for) -- this is what's shown
   * to the user instead of raw entry/stop/target numbers, which are still
   * computed above (needed for R/R gating and confluence) but no longer
   * surfaced directly in Telegram or the dashboard. */
  justification_ru: string | null;
  justification_en: string | null;
  used_fallback_levels: boolean;
  rs_pctile_estimate: number | null;
  rs_pctile_is_estimate: true;
  earnings_trading_days: number | null;
  options_wall_source: string | null;
}

function tradingDays(calendarDays: number): number {
  return (calendarDays * 5) / 7;
}

interface SectorThemeInfo {
  sector: string | null;
  beatsSpy: boolean;
  sectorRrgQuadrant: Quadrant | null;
  theme: string | null;
  themeConfidence: "exact" | "approximate" | null;
  themeRrgQuadrant: Quadrant | null;
}

async function sectorAndThemeInfo(ticker: string): Promise<SectorThemeInfo> {
  const [{ sector: yfSector, industry }, sectors] = await Promise.all([
    fetchSectorAndIndustry(ticker), scoreSectors(),
  ]);
  const mapped = yfSector ? (YF_SECTOR_MAP[yfSector] ?? yfSector) : null;
  const row = mapped ? sectors.find((r) => r.sector === mapped) : undefined;
  const beatsSpy = row ? row["4W_vs_SPY"] > 0 : false;

  const themeMatch = mapToTheme(ticker, industry);

  const [sectorRrg, themeRrg] = await Promise.all([
    mapped ? getLatestRrgReading(mapped) : Promise.resolve(null),
    themeMatch ? getLatestRrgReading(themeMatch.theme) : Promise.resolve(null),
  ]);

  return {
    sector: mapped, beatsSpy, sectorRrgQuadrant: sectorRrg?.quadrant ?? null,
    theme: themeMatch?.theme ?? null, themeConfidence: themeMatch?.confidence ?? null,
    themeRrgQuadrant: themeRrg?.quadrant ?? null,
  };
}

async function rsPctileEstimate(t: TechRead): Promise<number | null> {
  if (t.rs_raw == null) return null;
  const spyBars = await fetchDailyBars("SPY", "2y");
  if (spyBars.length === 0) return null;
  const spyRs = rsWeightedReturn(spyBars.map((b) => b.close));
  if (spyRs == null) return null;
  const margin = t.rs_raw - spyRs;
  const pctile = 50 + (margin / RS_PROXY_SPY_MARGIN_FOR_85TH) * 35;
  return Math.round(Math.min(Math.max(pctile, 0), 100) * 10) / 10;
}

function aplusChecklist(
  regimeScore: number, sectorBeatsSpy: boolean, t: TechRead, vision: VisionGrade,
  rr: number | null, rsPctile: number | null, earningsTradingDays: number | null
): { score: number; detail: { check: string; ok: boolean }[] } {
  const checks: [string, boolean][] = [
    ["Regime 3/4 or 4/4", regimeScore >= 3],
    ["Sector beats SPY (4W)", sectorBeatsSpy],
    ["RS percentile >= 85 (estimate)", (rsPctile ?? 0) >= 85],
    ["Recognizable pattern", !!vision.pattern && vision.pattern !== "none"],
    ["VDU present", !!(t.vdu || vision.vdu)],
    ["Clear pivot", t.pivot !== null && !t.extended],
    ["Logical stop", true],
    ["R/R >= 2:1", rr !== null && rr >= 2.0],
    [`Earnings >= ${MIN_EARNINGS_TRADING_DAYS} trading days away`,
      earningsTradingDays === null || earningsTradingDays >= MIN_EARNINGS_TRADING_DAYS],
  ];
  return { score: checks.filter(([, ok]) => ok).length, detail: checks.map(([check, ok]) => ({ check, ok })) };
}

function detectTrigger(t: TechRead, vision: VisionGrade): "breakout" | "pullback" | null {
  if (t.pivot !== null && t.dist_to_pivot_pct !== null) {
    if (t.dist_to_pivot_pct >= -1.0 && !t.extended && t.vol_ratio >= BREAKOUT_RVOL_MIN) return "breakout";
  }
  if (t.vdu || vision.vdu) return "pullback";
  return null;
}

export async function analyzeTicker(
  rawTicker: string,
  optionsToken: string | null = null,
  gradeFn: typeof gradeChart = gradeChart // injectable for tests -- avoids needing a live OpenAI call
): Promise<Verdict> {
  const ticker = rawTicker.toUpperCase().trim();
  const bars = await fetchDailyBars(ticker, "2y");
  const base: Omit<Verdict, "verdict" | "reason"> = {
    ticker, conviction: null, rr_band: null, trigger: null, aplus_score: 0, aplus_detail: [],
    regime_score: 0, regime_mode: "", sector: null, sector_beats_spy: false,
    sector_rrg_quadrant: null, theme: null, theme_match_confidence: null, theme_rrg_quadrant: null,
    price: null, entry: null, stop: null, target: null, rr: null,
    confluence_count: 0, confluence_signals: [], chart_grade: null, vision_note: null,
    justification_ru: null, justification_en: null,
    used_fallback_levels: false,
    rs_pctile_estimate: null, rs_pctile_is_estimate: true, earnings_trading_days: null,
    options_wall_source: null,
  };
  if (bars.length === 0) return { ...base, verdict: "PASS", reason: "no price data" };

  const t = readTechnicals(bars);
  const [regime, sectorTheme, earningsCalDays, rsPctile] = await Promise.all([
    assessRegime(),
    sectorAndThemeInfo(ticker),
    fetchNextEarningsCalendarDays(ticker),
    rsPctileEstimate(t),
  ]);
  const earningsTradingDays = earningsCalDays !== null ? tradingDays(earningsCalDays) : null;

  // Render the real chart before grading -- a rendering failure degrades to
  // a null image (visionGrade.ts falls back to a more-conservative
  // numbers-only prompt) rather than blocking the whole analysis.
  let chartImageDataUri: string | null = null;
  try {
    chartImageDataUri = await renderChartPng(ticker, bars, t);
  } catch (e) {
    console.error(`renderChartPng(${ticker}) failed: ${e instanceof Error ? e.message : e}`);
  }

  const vision = await gradeFn(ticker, chartImageDataUri, {
    price: t.price, ema9: t.ema9, ema21: t.ema21, ema50: t.ema50, ema200: t.ema200,
    rsi14: t.rsi14, atr_pct: t.atr_pct, pivot: t.pivot, dist_to_pivot_pct: t.dist_to_pivot_pct,
    vol_ratio: t.vol_ratio, vdu: t.vdu, fib_382: t.fib_382, fib_500: t.fib_500, fib_618: t.fib_618,
    vwap_anchor: t.vwap_anchor, poc: t.poc, val: t.val, vah: t.vah, sr_zones: t.sr_zones,
    recent_bars: bars.slice(-12),
  });
  const chartGrade = vision.grade || "C";

  // The numbers-only grading call sometimes declines to give a plan
  // (entry_type "none" or missing stop/target) where the real chart-vision
  // pipeline would have judged real structural levels. Rather than silently
  // discard the ticker as "R/R n/a", fall back to the same formulaic
  // stop_and_entry() the daily pipeline itself falls back to when it can't
  // do better -- always labeled as a fallback, never presented as the
  // model's own judgment.
  let entry = vision.entry ?? t.price;
  let stop = vision.stop ?? null;
  let target = vision.target ?? null;
  let usedFallbackLevels = false;
  if (stop === null || target === null) {
    const fallback = stopAndEntry(t);
    entry = vision.entry ?? fallback.entry;
    stop = fallback.stop;
    target = fallback.target;
    usedFallbackLevels = true;
  }
  const rr = stop && target && entry > stop ? Math.round(((target - entry) / (entry - stop)) * 100) / 100 : null;

  const walls: OptionsWalls = await readOptionsWalls(ticker, t.price, optionsToken);
  const { count: confCount, signals: confSignals } = confluenceCount(t, entry, walls);

  const v: Verdict = {
    ...base, verdict: "WAIT", reason: "",
    regime_score: regime.score, regime_mode: regime.mode,
    sector: sectorTheme.sector, sector_beats_spy: sectorTheme.beatsSpy,
    sector_rrg_quadrant: sectorTheme.sectorRrgQuadrant,
    theme: sectorTheme.theme, theme_match_confidence: sectorTheme.themeConfidence,
    theme_rrg_quadrant: sectorTheme.themeRrgQuadrant,
    price: t.price, entry, stop, target, rr,
    confluence_count: confCount, confluence_signals: confSignals,
    chart_grade: chartGrade, vision_note: vision.note ?? null,
    justification_ru: vision.justification_ru ?? null, justification_en: vision.justification_en ?? null,
    used_fallback_levels: usedFallbackLevels,
    rs_pctile_estimate: rsPctile, earnings_trading_days: earningsTradingDays,
    options_wall_source: walls.source,
  };

  // ---- PASS gates ----
  if (!t.above_200) return { ...v, verdict: "PASS", reason: "below EMA200 -- no long structure" };
  if (rr === null || rr < MIN_RR) {
    const why = usedFallbackLevels
      ? `structural stop/target (fallback -- grading gave no plan${vision.note ? `: "${vision.note}"` : ""})`
      : "grading's own entry/stop/target";
    return { ...v, verdict: "PASS", reason: `R/R ${rr ?? "n/a"} < ${MIN_RR} from ${why}` };
  }
  if (chartGrade === "F") return { ...v, verdict: "PASS", reason: `chart grade F -- ${vision.note ?? ""}` };
  if (earningsTradingDays !== null && earningsTradingDays < MIN_EARNINGS_TRADING_DAYS) {
    return { ...v, verdict: "PASS", reason: `earnings in ~${earningsTradingDays.toFixed(0)} trading days` };
  }
  if (vision.base_number != null && vision.base_number >= MAX_BASE_NUMBER) {
    return { ...v, verdict: "PASS", reason: `base #${vision.base_number} -- too extended a series of bases` };
  }

  const { score: aplusScore, detail: aplusDetail } = aplusChecklist(
    regime.score, sectorTheme.beatsSpy, t, vision, rr, rsPctile, earningsTradingDays
  );
  v.aplus_score = aplusScore;
  v.aplus_detail = aplusDetail;
  const trigger = detectTrigger(t, vision);
  v.trigger = trigger;

  // ---- WAIT caps ----
  if (regime.score <= 1) return { ...v, verdict: "WAIT", reason: "regime NO_TRADE -- watchlist only" };
  if (aplusScore <= APLUS_WAIT_CEILING) return { ...v, verdict: "WAIT", reason: `A+ score ${aplusScore}/9 -- no entry` };
  if (t.extended) return { ...v, verdict: "WAIT", reason: `${t.dist_to_pivot_pct!.toFixed(1)}% past pivot -- chase risk` };
  if (confCount === 0) return { ...v, verdict: "WAIT", reason: "no confluence at proposed entry" };
  if (trigger === null) return { ...v, verdict: "WAIT", reason: "structurally fine, no trigger today" };

  // ---- BUY ----
  // Conviction tiers: "A+" is the full 9/9 checklist (unchanged, matches
  // STRATEGY.md's canonical A+ checklist). Below that, a sector or theme
  // sitting in the RRG Improving/Leading quadrant is a real, separate
  // tailwind signal -- same idea as the daily pipeline treating sector-
  // rotation stage (Building/Emerging/Leading/Fading) as its own dimension
  // alongside the A+ score, not fused into the 9 questions themselves. This
  // is the one place RRG data affects the verdict; everywhere else it's
  // shown as context only.
  const rotationTailwind = (q: Quadrant | null) => q === "Improving" || q === "Leading";
  const hasRotationTailwind = rotationTailwind(sectorTheme.sectorRrgQuadrant) || rotationTailwind(sectorTheme.themeRrgQuadrant);
  const conviction: "A+" | "A" | "standard" =
    aplusScore === 9 ? "A+" : hasRotationTailwind ? "A" : "standard";
  const rrBand: "2:1-2.9:1" | "3:1+" = rr >= 3.0 ? "3:1+" : "2:1-2.9:1";
  return {
    ...v, verdict: "BUY", conviction, rr_band: rrBand,
    reason: `${trigger} trigger, A+ ${aplusScore}/9, R/R ${rr.toFixed(1)}`,
  };
}
